

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genopandas.ngs.cnv &mdash; genopandas 0.1.dev30+g62366b5.d20180224 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="genopandas 0.1.dev30+g62366b5.d20180224 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> genopandas
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">genopandas</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>genopandas.ngs.cnv</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genopandas.ngs.cnv</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">genopandas.core.frame</span> <span class="k">import</span> <span class="n">GenomicDataFrame</span>
<span class="kn">from</span> <span class="nn">genopandas.core.matrix</span> <span class="k">import</span> <span class="n">AnnotatedMatrix</span><span class="p">,</span> <span class="n">GenomicMatrix</span>


<div class="viewcode-block" id="CnvValueMatrix"><a class="viewcode-back" href="../../../api.html#genopandas.ngs.cnv.CnvValueMatrix">[docs]</a><span class="k">class</span> <span class="nc">CnvValueMatrix</span><span class="p">(</span><span class="n">GenomicMatrix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;CnvMatrix containing (segmented) logratio values (positions-by-samples).</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="CnvValueMatrix.as_segments"><a class="viewcode-back" href="../../../api.html#genopandas.ngs.cnv.CnvValueMatrix.as_segments">[docs]</a>    <span class="k">def</span> <span class="nf">as_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">drop_columns</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns matrix as segments (consecutive stetches with same value).</span>

<span class="sd">        Assumes that values have already been segmented, i.e. that bins</span>
<span class="sd">        in the same segment have been assigned same numeric value.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        drop_columns : bool</span>
<span class="sd">            Whether to drop chromosome, start, end and sample columns</span>
<span class="sd">            after setting the index.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        GenomicDataFrame</span>
<span class="sd">            GenomicDataFrame describing genomic segments. Indexed by</span>
<span class="sd">            chromosome, start, end and sample.</span>

<span class="sd">            Note that the sample index is included to avoid duplicate index</span>
<span class="sd">            errors when reindexing in cases where samples have identical</span>
<span class="sd">            segments.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>

        <span class="c1"># Get segments per sample.</span>
        <span class="n">segment_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_segments_for_sample</span><span class="p">(</span><span class="n">sample_values</span><span class="p">)</span>
             <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">sample_values</span> <span class="ow">in</span> <span class="n">values</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># yapf: disable</span>

        <span class="c1"># Set index. Note that we add sample here, to avoid running into</span>
        <span class="c1"># duplicate index errors when re-indexing later down the road.</span>
        <span class="c1"># This is a bit of a hack, but GenomicDataFrames shouldn&#39;t suffer</span>
        <span class="c1"># from having any extra index entries.</span>
        <span class="n">segment_data</span> <span class="o">=</span> <span class="n">segment_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span>
            <span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">],</span> <span class="n">drop</span><span class="o">=</span><span class="n">drop_columns</span><span class="p">)</span>

        <span class="n">segments</span> <span class="o">=</span> <span class="n">GenomicDataFrame</span><span class="p">(</span><span class="n">segment_data</span><span class="p">)</span>
        <span class="n">segments</span> <span class="o">=</span> <span class="n">segments</span><span class="o">.</span><span class="n">gloc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gloc</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">segments</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_segments_for_sample</span><span class="p">(</span><span class="n">sample_values</span><span class="p">):</span>
        <span class="c1"># Calculate segment ids (distinguished by diff values).</span>
        <span class="n">segment_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">_padded_diff</span><span class="p">(</span><span class="n">sample_values</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Get sample and position columns.</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">sample_values</span><span class="o">.</span><span class="n">name</span>
        <span class="n">chrom_col</span><span class="p">,</span> <span class="n">start_col</span><span class="p">,</span> <span class="n">end_col</span> <span class="o">=</span> <span class="n">sample_values</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span>

        <span class="c1"># Group and determine positions + values.</span>
        <span class="n">grouped</span> <span class="o">=</span> <span class="n">sample_values</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="n">chrom_col</span><span class="p">,</span> <span class="n">segment_ids</span><span class="p">])</span>

        <span class="n">segments</span> <span class="o">=</span> <span class="n">grouped</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
            <span class="n">chrom_col</span><span class="p">:</span> <span class="s1">&#39;first&#39;</span><span class="p">,</span>
            <span class="n">start_col</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span>
            <span class="n">end_col</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span>
            <span class="n">sample</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="s1">&#39;size&#39;</span><span class="p">]</span>
        <span class="p">})</span>

        <span class="c1"># Flatten column levels and rename.</span>
        <span class="n">segments</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">segments</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">segments</span> <span class="o">=</span> <span class="n">segments</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span>
            <span class="n">chrom_col</span> <span class="o">+</span> <span class="s1">&#39;_first&#39;</span><span class="p">:</span> <span class="s1">&#39;chromosome&#39;</span><span class="p">,</span>
            <span class="n">start_col</span> <span class="o">+</span> <span class="s1">&#39;_min&#39;</span><span class="p">:</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span>
            <span class="n">end_col</span> <span class="o">+</span> <span class="s1">&#39;_max&#39;</span><span class="p">:</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span>
            <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_first&#39;</span><span class="p">:</span> <span class="s1">&#39;value&#39;</span><span class="p">,</span>
            <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_size&#39;</span><span class="p">:</span> <span class="s1">&#39;size&#39;</span>
        <span class="p">})</span>

        <span class="c1"># Add sample name and reorder columns.</span>
        <span class="n">segments</span> <span class="o">=</span> <span class="n">segments</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;value&#39;</span><span class="p">,</span> <span class="s1">&#39;size&#39;</span><span class="p">])</span>
        <span class="n">segments</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample</span>

        <span class="k">return</span> <span class="n">segments</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="CnvValueMatrix.to_igv"><a class="viewcode-back" href="../../../api.html#genopandas.ngs.cnv.CnvValueMatrix.to_igv">[docs]</a>    <span class="k">def</span> <span class="nf">to_igv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Saves data for viewing in IGV.&quot;&quot;&quot;</span>

        <span class="n">igv_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

        <span class="c1"># Rename index columns.</span>
        <span class="n">igv_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;Start&#39;</span><span class="p">,</span> <span class="s1">&#39;End&#39;</span><span class="p">]</span>
        <span class="n">column_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span><span class="p">,</span> <span class="n">igv_columns</span><span class="p">))</span>

        <span class="n">igv_data</span> <span class="o">=</span> <span class="n">igv_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">column_map</span><span class="p">)</span>

        <span class="c1"># Add &#39;Feature&#39; column.</span>
        <span class="n">feature_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;P</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">igv_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
        <span class="n">igv_data</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="s1">&#39;Feature&#39;</span><span class="p">,</span> <span class="n">feature_names</span><span class="p">)</span>

        <span class="c1"># Write file.</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;#type=COPY_NUMBER&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">file_</span><span class="p">)</span>
            <span class="n">igv_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">file_</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="CnvCallMatrix"><a class="viewcode-back" href="../../../api.html#genopandas.ngs.cnv.CnvCallMatrix">[docs]</a><span class="k">class</span> <span class="nc">CnvCallMatrix</span><span class="p">(</span><span class="n">AnnotatedMatrix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Cnv matrix containing CNV calls (genes-by-samples).&quot;&quot;&quot;</span>

<div class="viewcode-block" id="CnvCallMatrix.mask_with_controls"><a class="viewcode-back" href="../../../api.html#genopandas.ngs.cnv.CnvCallMatrix.mask_with_controls">[docs]</a>    <span class="k">def</span> <span class="nf">mask_with_controls</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">column</span><span class="p">,</span> <span class="n">mask_value</span><span class="o">=</span><span class="mf">0.0</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Masks calls present in control samples.</span>

<span class="sd">        Calls are retained if (a) no call is present in the matched control</span>
<span class="sd">        sample, (b) if the sample call is more extreme than the control sample</span>
<span class="sd">        or (c) the sample and control have calls with different signs</span>
<span class="sd">        (loss/gain).</span>

<span class="sd">        Matched control samples should be indicated by the given column</span>
<span class="sd">        in the sample_data annotation.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">control_samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_data</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

        <span class="n">new_values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">ctrl</span> <span class="ow">in</span> <span class="nb">dict</span><span class="p">(</span><span class="n">control_samples</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_call_mask</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="p">[</span><span class="n">ctrl</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="p">[</span><span class="n">sample</span><span class="p">])</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">mask</span><span class="p">,</span> <span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">mask_value</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_constructor</span><span class="p">(</span><span class="n">new_values</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_call_mask</span><span class="p">(</span><span class="n">ctrl_values</span><span class="p">,</span> <span class="n">sample_values</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns mask in which entries are True where ctrl and sample</span>
<span class="sd">           have different signs or the sample has a more extreme value.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ctrl_sign</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">(</span><span class="n">ctrl_values</span><span class="p">)</span>
        <span class="n">sample_sign</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">(</span><span class="n">sample_values</span><span class="p">)</span>

        <span class="n">diff_sign</span> <span class="o">=</span> <span class="p">(</span><span class="n">ctrl_sign</span> <span class="o">-</span> <span class="n">sample_sign</span><span class="p">)</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mf">1e-8</span>
        <span class="n">higher_val</span> <span class="o">=</span> <span class="n">sample_values</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;</span> <span class="n">ctrl_values</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">diff_sign</span> <span class="o">|</span> <span class="p">(</span><span class="o">~</span><span class="n">diff_sign</span> <span class="o">&amp;</span> <span class="n">higher_val</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">_padded_diff</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">pad_value</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Same as np.diff, with leading 0 to keep same length as input.&quot;&quot;&quot;</span>
    <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">pad</span><span class="p">(</span>
        <span class="n">diff</span><span class="p">,</span> <span class="n">pad_width</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;constant&#39;</span><span class="p">,</span> <span class="n">constant_values</span><span class="o">=</span><span class="n">pad_value</span><span class="p">)</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Julian de Ruiter.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1.dev30+g62366b5.d20180224',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>